File: 5thSem/AIES/Assignment1.py
import heapq

n = 3

dirOptions = [(1,0), (-1, 0), (0, 1), (0, -1)]
dirChar = ['D', 'U', 'R', 'L']

class Node :
    def __init__(self, state, g, h, path, blankPos):
        self.state = state
        self.g = g
        self.h = h
        self.path = path
        self.blankPos = blankPos

    def f(self) :
        return self.g + self.h
    
    def __lt__(self, other) :
        return self.f() < other.f()
    
def calcHeur(state, goal) :
    goalPos = {}
    for i in range(n):
        for j in range(n) :
            goalPos[goal[i][j]] = (i, j)
    
    dist = 0
    for i in range(n):
        for j in range(n) :
            val = state[i][j]
            if(val != 0) :
                gi, gj = goalPos[val]
                dist += abs(i-gi) + abs(j-gj) #Manhattan Distance
    return dist

def matToString(state) :
    num = ""
    for i in range(n) :
        for j in range(n) :
            num += str(state[i][j])
    return num

def isValid(x, y) :
    return 0 <= x < n and 0 <= y < n

def display(state) :
    for row in state:
        print(row)
    print()

def aStar(start, goal) :
    h = calcHeur(start, goal)
    blank = None #
    for i in range(n):
        for j in range(n) :
            if(start[i][j] == 0):
                blank = (i,j)
    openList = []
    closedList = set()
    startNode = Node(start, 0, h, "", blank)
    heapq.heappush(openList, startNode) #

    while openList:
        curr = heapq.heappop(openList)
        stateId = matToString(curr.state)

        if stateId in closedList :
            continue #
        closedList.add(stateId) #

        if curr.state == goal :
            print(f"Solution found in {curr.g} moves!")
            print(f"Path : {curr.path}")
            display(curr.state)
            return True
        
        for d in range(4):
            newX = curr.blankPos[0] + dirOptions[d][0]
            newY = curr.blankPos[1] + dirOptions[d][1]

            if isValid(newX, newY):
                newState = [row[:] for row in curr.state]
                newState[curr.blankPos[0]][curr.blankPos[1]], newState[newX][newY] = newState[newX][newY], newState[curr.blankPos[0]][curr.blankPos[1]]

                newId = matToString(newState)

                if newId not in closedList:
                    newG = curr.g + 1
                    newH = calcHeur(newState, goal)
                    newPath = curr.path + dirChar[d]
                    heapq.heappush(openList, Node(newState, newG, newH, newPath, (newX, newY)))
    print("No solution found")
    return False

def inputMatrix():
    mat = []
    print("Enter numbers row by row: ")
    print()
    for _ in range(n):
        row = list(map(int, input().split()))
        mat.append(row)
    return mat

def main():
    print("Initial matrix")
    start = [[1,2,3],
             [4,5,6],
             [7,8,0]]
    display(start)
    print("Goal matrix")
    goal =  [[0,1,2],
             [3,4,5],
             [6,7,8]]
    display(goal)
    aStar(start, goal)

if __name__ == "__main__":
    main()

File: 5thSem/AIES/Assignment1.cpp
#include <iostream>
#include <vector>
#include <queue>
#include <unordered_set>
#include <cmath>
#include <map>

using namespace std;

const int N = 3;
typedef vector<vector<int>> Matrix;

struct Node {
    Matrix state;
    int g, h;
    string path;
    pair<int, int> blank_pos;
    
    Node(Matrix st, int g_cost, int h_cost, string p, pair<int, int> bp)
        : state(st), g(g_cost), h(h_cost), path(p), blank_pos(bp) {}

    int f() const { return g + h; }

    bool operator>(const Node &other) const {
        return f() > other.f();
    }
};

// Manhattan distance
int calculateHeuristic(const Matrix& state, const Matrix& goal) {
    map<int, pair<int, int>> goalPos;
    for (int i = 0; i < N; ++i)
        for (int j = 0; j < N; ++j)
            goalPos[goal[i][j]] = {i, j};

    int dist = 0;
    for (int i = 0; i < N; ++i)
        for (int j = 0; j < N; ++j)
            if (state[i][j] != 0)
                dist += abs(i - goalPos[state[i][j]].first) + abs(j - goalPos[state[i][j]].second);
    return dist;
}

string matrixToString(const Matrix& state) {
    string s;
    for (const auto& row : state)
        for (int val : row)
            s += to_string(val);
    return s;
}

vector<pair<int, int>> directions = {{-1,0}, {1,0}, {0,-1}, {0,1}};
vector<char> dirChar = {'U', 'D', 'L', 'R'};

bool isValid(int x, int y) {
    return x >= 0 && y >= 0 && x < N && y < N;
}

void printMatrix(const Matrix& m) {
    for (auto& row : m) {
        for (int val : row)
            cout << val << " ";
        cout << "\n";
    }
    cout << "\n";
}

bool AStar(Matrix start, Matrix goal) {
    auto h = calculateHeuristic(start, goal);
    pair<int, int> blank;
    for (int i = 0; i < N; ++i)
        for (int j = 0; j < N; ++j)
            if (start[i][j] == 0)
                blank = {i, j};

    priority_queue<Node, vector<Node>, greater<Node>> open;
    unordered_set<string> closed;
    open.push(Node(start, 0, h, "", blank));

    while (!open.empty()) {
        Node curr = open.top();
        open.pop();

        string id = matrixToString(curr.state);
        if (closed.find(id) != closed.end())
            continue;

        closed.insert(id);

        if (curr.state == goal) {
            cout << "Solution found in " << curr.g << " moves!\nPath: " << curr.path << "\n\n";
            cout << "Final state:\n";
            printMatrix(curr.state);
            return true;
        }

        for (int d = 0; d < 4; ++d) {
            int newX = curr.blank_pos.first + directions[d].first;
            int newY = curr.blank_pos.second + directions[d].second;

            if (isValid(newX, newY)) {
                Matrix newState = curr.state;
                swap(newState[curr.blank_pos.first][curr.blank_pos.second], newState[newX][newY]);
                string newID = matrixToString(newState);

                if (closed.find(newID) == closed.end()) {
                    int newG = curr.g + 1;
                    int newH = calculateHeuristic(newState, goal);
                    string newPath = curr.path + dirChar[d];
                    open.push(Node(newState, newG, newH, newPath, {newX, newY}));
                }
            }
        }
    }
    cout << "No solution found.\n";
    return false;
}

Matrix inputMatrix(string prompt) {
    Matrix mat(N, vector<int>(N));
    cout << "Enter " << prompt << " state (row by row, 0 for blank):\n";
    for (int i = 0; i < N; ++i)
        for (int j = 0; j < N; ++j)
            cin >> mat[i][j];
    return mat;
}

int main() {
    Matrix start = inputMatrix("initial");
    Matrix goal = inputMatrix("goal");

    AStar(start, goal);
    return 0;
}
